blockchain/chain/src/store/index.rs

```diff
@@ -16,6 +17,71 @@ const DEFAULT_CHAIN_INDEX_CACHE_SIZE: NonZeroUsize =
+pub(super) mod checkpoint_tipsets {
+    use cid::Cid;
+    use forest_blocks::TipsetKeys;
+    use once_cell::sync::Lazy;
+    use std::collections::HashMap;
+    use std::str::FromStr;
+
+    const CALIBNET_GENESIS_CID: &str =
+        "bafy2bzacecyaggy24wol5ruvs6qm73gjibs2l2iyhcqmvi7r7a4ph7zx3yqd4";
+    const MAINNET_GENESIS_CID: &str =
+        "bafy2bzacecnamqgqmifpluoeldx7zzglxcljo6oja4vrmtj7432rphldpdmm2";
+
+    macro_rules! add_calibnet {
+        ($map: ident, $key_hash:expr) => {
+            $map.insert(
+                $key_hash,
+                // calibnet genesis tipset keys
+                TipsetKeys::new(vec![Cid::from_str(CALIBNET_GENESIS_CID).unwrap()]),
+            );
+        };
+    }
+
+    macro_rules! add_mainnet {
+        ($map: ident, $key_hash:expr) => {
+            $map.insert(
+                $key_hash,
+                // mainnet genesis tipset keys
+                TipsetKeys::new(vec![Cid::from_str(MAINNET_GENESIS_CID).unwrap()]),
+            );
+        };
+    }
+
+    // The hashes for these checkpoints is obtained by passing the tipset keys' cids (tipset.cids())
+    // through a blake2b hasher.
+    type TipsetKeyHash = &'static str;
+    type GenesisTipsetCids = TipsetKeys;
+    // A map of validated checkpoints mapping to their genesis tipkeys
+    type TipsetCheckpointsRegistry = HashMap<TipsetKeyHash, GenesisTipsetCids>;
+    // Represents a static map of validated tipset hashes which helps to remove the need to validate the tipset
+    // back to genesis if it has been validated before, thereby reducing boot times.
+    pub(super) static TIPSET_CHECKPOINTS: Lazy<TipsetCheckpointsRegistry> = Lazy::new(|| {
+        let mut map = HashMap::new();
+        // The numeric suffix in the variable name is the height of the snapshot.
+        // NB: Add desired tipset checkpoints below this by using RPC command: forest-cli chain tipset-hash <cid keys>
+        const CALIBNET_CHECKPOINT_1405400: TipsetKeyHash = "7930ad8bf32b35314b3bc47b9e25249af8ec6ba7f5544c05e8b5bd3b3ec09f76df8bd2278f9b318badf1a08d0a468abd55130465c6c55f99e67badc0e614ca79";
+        const MAINNET_CHECKPOINT_2325300: TipsetKeyHash = "319f2351ceaf78fbcc8688dc75a19bdf8ee6e895e547ff5cc2f7b18a3a36b65ff94c1860733137d0244352f82ba6fd9672aec14deee358e7cf6e088bf89a28b1";
+        const CALIBNET_CHECKPOINT_41000: TipsetKeyHash = "1a11a07d427348cc14eaa901de1ba9c6a4e18400bb557f5a0fabbcb22352319e31a7bc988a92525339f84275c0ef6dfbffcb50bb9d9843701875eecfa3ccb069";
+        add_calibnet!(map, CALIBNET_CHECKPOINT_1405400);
+        add_calibnet!(map, CALIBNET_CHECKPOINT_41000);
+        add_mainnet!(map, MAINNET_CHECKPOINT_2325300);
+        map
+    });
+
+    pub(super) fn genesis_from_checkpoint_tipset(tsk: &TipsetKeys) -> Option<GenesisTipsetCids> {
+        TIPSET_CHECKPOINTS.get(&tipset_hash(tsk).as_str()).cloned()
+    }
+
+    pub fn tipset_hash(tsk: &TipsetKeys) -> String {
```

Comment:
Would it be beneficial not to operate on strings (unless displaying them to the user)? Operating on raw bytes is often cheaper, easier to compare, no UTF-8 wide chars in the back etc.

Reply:
> Always operate on raw bytes, never on encoded strings. Only use hex and base64 for pretty-printing.

from cryptopals :)

Reply:
I agree on using byte array would be perf friendly, but unfortunately we need to store these as constant values in chain_store/index.rs so a hash was ideal, compared to storing a 128 size byte array and the tipset_hash function is used in the rpc call made by the user.

Reply:
Also, tipset_hash will only get invoked during initial scan and only for snapshot loads and it likely won't run for long as the tipset checkpoint will be near the head of the snapshot.